File: problems/1733_minimum_number_of_people_to_teach/1733_minimum_number_of_people_to_teach.py
class Solution:
    def minimumTeachings(self, n: int, languages: List[List[int]], friendships: List[List[int]]) -> int:
        user_languages = [set(langs) for langs in languages]
        need_to_teach = set()
        for u, v in friendships:
            u -= 1
            v -= 1
            if user_languages[u].isdisjoint(user_languages[v]):
                need_to_teach.add(u)
                need_to_teach.add(v)
        language_count = defaultdict(int)
        for user in need_to_teach:
            for lang in user_languages[user]:
                language_count[lang] += 1
        max_known = max(language_count.values(), default=0)
        return len(need_to_teach) - max_known

File: problems/1733_minimum_number_of_people_to_teach/1733_minimum_number_of_people_to_teach.java
class Solution {
    public int minimumTeachings(int n, int[][] languages, int[][] friendships) {
        // Store each user knows languages
        Set<Integer>[] userLanguages = new Set[languages.length];
        for (int i = 0; i < languages.length; i++) {
            userLanguages[i] = new HashSet<>();
            for (int lang : languages[i]) {
                userLanguages[i].add(lang);
            }
        }
        // Record know need to teach language users
        Set<Integer> needToTeach = new HashSet<>();
        for (int[] f : friendships) {
            int u = f[0] - 1, v = f[1] - 1;
            Set<Integer> setU = userLanguages[u];
            Set<Integer> setV = userLanguages[v];
            if (Collections.disjoint(setU, setV)) { // Without common language
                needToTeach.add(u);
                needToTeach.add(v);
            }
        }
        // Count the number of people who know each language among those who need to be
        // taught
        int[] languageCount = new int[n + 1];
        for (int user : needToTeach) {
            for (int lang : userLanguages[user]) {
                languageCount[lang]++;
            }
        }
        int maxKnown = 0;
        for (int i = 1; i <= n; i++) {
            maxKnown = Math.max(maxKnown, languageCount[i]);
        }
        return needToTeach.size() - maxKnown;
    }
}